system/security/digisign/digisign.cpp

```diff
@@ -0,0 +1,255 @@
+            EVP_THROW("digiSign:EVP_MD_CTX_create: %s");
+        }
+
+        //initialize context for SHA-256 hashing function
+        int rc;
+        if (isSigning)
+            rc = EVP_DigestSignInit(RSACtx, nullptr, EVP_sha256(), nullptr, pKey);
+        else
+            rc = EVP_DigestVerifyInit(RSACtx, nullptr, EVP_sha256(), nullptr, pKey);
+        if (rc <= 0)
+        {
+            EVP_CLEANUP(pKey, RSACtx);//cleans allocated key and digest context
+            if (isSigning)
+                EVP_THROW("digiSign:EVP_DigestSignInit: %s")
+            else
+                EVP_THROW("digiSign:EVP_DigestVerifyInit: %s")
+        }
+        *ctx = RSACtx;
+        *PKey = pKey;
+        return true;
+    }
+
+    //Create base 64 encoded digital signature of given text string
+    bool digiSign(const char * text, StringBuffer & b64Signature)
+    {
+        if (!signingConfigured)
+            throw MakeStringException(-1, "digiSign:Creating Digital Signatures not configured");
+
+        if (privateKeyBuff.isEmpty())
+            privateKeyBuff.loadFile(privateKeyFile.str());
+        if (privateKeyBuff.isEmpty())
+            throw MakeStringException(-1, "digiSign:Cannot load private key file");
+
+        CriticalBlock b(digiSignCrit);
+
+#ifdef _USE_OPENSSL
+        EVP_MD_CTX * RSACtx;
+        EVP_PKEY * EVPKey;
+        digiInit(true, privateKeyBuff.str(), passphraseBuff.str(), &RSACtx, &EVPKey);
+
+        //add string to the context
+        if (EVP_DigestSignUpdate(RSACtx, (size_t*)text, strlen(text)) <= 0)
+        {
+            EVP_CLEANUP(EVPKey, RSACtx);
+            EVP_THROW("digiSign:EVP_DigestSignUpdate: %s");
+        }
+
+        //compute length of signature
+        size_t encMsgLen;
+        if (EVP_DigestSignFinal(RSACtx, nullptr, &encMsgLen) <= 0)
+        {
+            EVP_CLEANUP(EVPKey, RSACtx);
+            EVP_THROW("digiSign:EVP_DigestSignFinal1: %s");
+        }
+
+        if (encMsgLen == 0)
+        {
+            EVP_CLEANUP(EVPKey, RSACtx);
+            EVP_THROW("digiSign:EVP_DigestSignFinal length returned 0: %s");
+        }
+
+        //compute signature (signed digest)
+        unsigned char * encMsg = (unsigned char*) malloc(encMsgLen);
+        if (encMsg == nullptr)
+        {
+            EVP_CLEANUP(EVPKey, RSACtx);
+            throw MakeStringException(-1, "digiSign:malloc(%u) returned NULL",(unsigned)encMsgLen);
+        }
+
+        if (EVP_DigestSignFinal(RSACtx, encMsg, &encMsgLen) <= 0)
+        {
+            free(encMsg);
+            EVP_CLEANUP(EVPKey, RSACtx);
+            EVP_THROW("digiSign:EVP_DigestSignFinal2: %s");
+        }
+        EVP_CLEANUP(EVPKey, RSACtx);
+
+        //convert to base64
+        JBASE64_Encode(encMsg, encMsgLen, b64Signature, false);
+
+        //cleanup
+        free(encMsg);
+        return true;//success
+#else
+        throw MakeStringException(-1, "digiSign:Platform built without OPENSSL");
+        return false;
```

Comment:
some code checkers will complain about unreachable code, cleaner to remove the return.

Reply:
same for the following function